caddytls/ctmonitor.go

```
@@ -0,0 +1,308 @@
+	v.Add("expand", "cert")
+	encodedValue := v.Encode()
+	query = config.Query + "?" + encodedValue
+	return query
+}
+
+func putLatestID(currentID int, fileName string) error {
+	writeValue := strconv.Itoa(currentID)
+	return ioutil.WriteFile(fileName, []byte(writeValue), 0600)
+}
+
+func queryDomainName(domainName string, config QueryConfig, biggestID *int, retrievedCerts map[string]string) (retryAfter string, err error) {
+	var querySize int
+
+	for ok := true; ok; ok = querySize > 0 {
+		querySize, config.Index, retryAfter, err = getCertSpotterCerts(domainName, config, biggestID, retrievedCerts)
+		if err != nil {
+			return retryAfter, fmt.Errorf("error: %v while getting certs from certSpotter", err) // Finish error checking.
+		}
+	}
+	return
+}
+
+func startMonitoring(eventType caddy.EventName, eventInfo interface{}) error {
+	go monitorCerts()
+	return nil
+}
+
+func getCertSpotterCerts(domainName string, config QueryConfig, biggestID *int, retrievedCerts map[string]string) (numOfIssuanceObjects int, issuanceID int, retryAfter string, err error) {
+	// issuanceObjects consists of the issuanceObjects returned from CertSpotter.
+	issuanceID = config.Index
+	var (
+		issuanceObjects []CertSpotterResponse
+		certQuery       string
+	)
+	certQuery = prepQuery(domainName, config)
+
+	response, err := http.Get(certQuery)
+	if err != nil {
+		log.Printf("https get request failed on input %s \nError: %v", prepQuery(domainName, config), err)
+	}
+	defer response.Body.Close()
+
+	err = json.NewDecoder(response.Body).Decode(&issuanceObjects) // handle error
+	if err != nil {
+		retryAfter = "3600"
+		return numOfIssuanceObjects, issuanceID, retryAfter, fmt.Errorf("decoding json stream: %v", err)
+	}
+
+	numOfIssuanceObjects = len(issuanceObjects)
+	if numOfIssuanceObjects > 0 {
+		for i, issuance := range issuanceObjects {
+			bytes, err := base64.StdEncoding.DecodeString(issuance.Cert.Data)
+			if err != nil {
+				log.Printf("Decoding failed: %v", err)
```

The body of this loop should probably be a function, then you can just return this error immediately; don't continue if this decode fails.